src/sql-server-util/src/lib.rs

```diff
@@ -0,0 +1,522 @@
+/// introduce a [`Transaction`] type that when dropped will cause the `TRANSACTION` in the
+/// connected SQL Server instance to get rolled back.
+#[derive(Debug)]
+pub struct Client {
+    tx: UnboundedSender<Request>,
+}
+// While a Client could implement Clone, it's not obvious how multiple Clients
+// using the same SQL Server connection would interact, so ban it for now.
+static_assertions::assert_not_impl_all!(Client: Clone);
+
+impl Client {
+    pub async fn connect(config: tiberius::Config) -> Result<(Self, Connection), anyhow::Error> {
+        let tcp = TcpStream::connect(config.get_addr()).await?;
+        tcp.set_nodelay(true)?;
+        Self::connect_raw(config, tcp).await
+    }
+
+    pub async fn connect_raw(
+        config: tiberius::Config,
+        tcp: tokio::net::TcpStream,
+    ) -> Result<(Self, Connection), anyhow::Error> {
+        let client = tiberius::Client::connect(config, tcp.compat_write())
+            .await
+            .context("connecting to SQL Server")?;
+        let (tx, rx) = tokio::sync::mpsc::unbounded_channel();
+
+        Ok((Client { tx }, Connection { rx, client }))
+    }
+
+    /// Executes SQL statements in SQL Server, returning the number of rows effected.
+    ///
+    /// Passthrough method for [`tiberius::Client::execute`].
+    ///
+    /// Note: The returned [`Future`] does not need to be awaited for the query
+    /// to be sent.
+    ///
+    /// [`Future`]: std::future::Future
+    pub async fn execute<'a>(
+        &mut self,
+        query: impl Into<Cow<'a, str>>,
+        params: &[&dyn ToSql],
+    ) -> Result<u64, anyhow::Error> {
+        let (tx, rx) = tokio::sync::oneshot::channel();
+
+        let params = params
+            .iter()
+            .map(|p| OwnedColumnData::from(p.to_sql()))
+            .collect();
+        let kind = RequestKind::Execute {
+            query: query.into().to_string(),
+            params,
+        };
+        self.tx
+            .send(Request { tx, kind })
+            .context("sending request")?;
+
+        let response = rx.await.context("channel")?.context("execute")?;
+        match response {
+            Response::Execute { rows_affected } => Ok(rows_affected),
+            other @ Response::Rows(_) => {
+                let err =
+                    anyhow::anyhow!("programming error! expected Response::Execute, got {other:?}");
+                Err(err)
+            }
+        }
+    }
+
+    /// Executes SQL statements in SQL Server, returning the resulting rows.
+    ///
+    /// Passthrough method for [`tiberius::Client::query`].
+    ///
+    /// Note: The returned [`Future`] does not need to be awaited for the query
+    /// to be sent.
+    ///
+    /// [`Future`]: std::future::Future
+    pub async fn query<'a, 'q>(
+        &'a mut self,
+        query: impl Into<Cow<'q, str>>,
+        params: &'q [&'q dyn tiberius::ToSql],
+    ) -> Result<SmallVec<[tiberius::Row; 1]>, anyhow::Error>
+    where
+        'a: 'q,
```

Comment:
We probably don't need `'q` given we're converting params into owned data anyways

Reply:
Fixed!